### PR TITLE
Util optimizations

### PR DESCRIPTION
**Pull request by Wicpar**
- removed lots of unchecked function with negligible performance impact (+-2%)
- improved truncate_str 50% to 90%
- made a safe version of `encode_char_utf8_unchecked` called `append_char_truncate` that operates on &mut [u8]

best to verify benches, it could be platform specific

### CODE REVIEW

**Comment by Wicpar** [comment]
well it looks like the benches are fake...

the const &'static str are inlied into the functions thus giving those crazy push times.
we should make sure the parameters are not inlined.

**Comment by Wicpar** [comment]
so try_from() is excessively slow in every case, unchecked is indeed the same speed except in the smallstring case where it is 12% faster. if 0.2 ns can be sacrificed in the small_string case it should be removed in favor of the safe ones

File: src/utils.rs [hunk 1]
```diff
@@ -156,17 +156,22 @@ pub fn is_char_boundary<const N: usize>(s: &ArrayString<N>, idx: u8) -> Result<(
...
-        unsafe { slice.get_unchecked(..size.into()) }
-    } else if (size as usize) < slice.len() {
-        let mut index = size.saturating_sub(1) as usize;
-        while !slice.is_char_boundary(index) {
-            index = index.saturating_sub(1);
-        }
-        unsafe { slice.get_unchecked(..index) }
-    } else {
-        slice
+    let mut size = size as usize;
```
Comment (paulocsanz) [inline src/utils.rs:159]: I like this idea that we only need to subtract at most 4 bytes, but I'm not sure if manually unrolling the loop is great. Shouldn't the compiler be able to optimize a simple for?

If not then sure, let's unroll it manually.

File: src/utils.rs [hunk 2]
```diff
@@ -138,8 +139,7 @@ where
     L: Into<usize>,
 {
     let (s, l) = (size.into(), limit.into());
-    trace!("Out of bounds: ensures {} < {}", s, l);
-    Some(()).filter(|_| s <= l).ok_or(OutOfBounds)
+    (s <= l).then_some(()).ok_or(OutOfBounds)
```
Comment (paulocsanz) [inline src/utils.rs:142]: Why remove the log? then_some is awesome btw!

File: src/arraystring.rs [hunk 3]
```diff
@@ -1,11 +1,12 @@
...
 
-use crate::utils::{encode_char_utf8_unchecked, is_char_boundary, is_inside_boundary, never};
+use crate::utils::{append_char_truncate, encode_char_utf8_unchecked, is_char_boundary, is_inside_boundary, never};
 use crate::utils::{shift_left_unchecked, shift_right_unchecked, truncate_str, IntoLossy};
 use crate::{error::Error, prelude::*};
 use core::char::{decode_utf16, REPLACEMENT_CHARACTER};
-use core::str::{from_utf8, from_utf8_unchecked};
+use core::str::from_utf8;
 use core::{cmp::min, ops::*, ptr::copy_nonoverlapping};
+use std::str::{from_utf8_unchecked, FromStr};
```
Comment (paulocsanz) [inline src/arraystring.rs:9]: std is an optional feature, so FromStr should only be implemented in that case. And for optional features we should be inlining the import in the usage.

Why did you move from_utf8_unchecked import for this?

File: src/arraystring.rs [hunk 4]
```diff
@@ -46,6 +47,16 @@ impl<const N: usize> ArrayString<N> {
         }
     }
 
+    #[inline]
+    fn remaining_mut(&mut self) -> &mut [u8] {
+        &mut self.array[self.size as usize..]
```
Comment (paulocsanz) [inline src/arraystring.rs:52]: We should not give access to invalid values in safe functions. If it's out of the valid string (0..len) it should only be accessible through unsafe functions.

Specially because we might use MaybeUninit now that is exists (at the time mem::uninitialized was too brittle)

File: src/arraystring.rs [hunk 5]
```diff
@@ -65,8 +76,8 @@ impl<const N: usize> ArrayString<N> {
     /// ```
     #[inline]
     pub fn try_from_str<S>(s: S) -> Result<Self, OutOfBounds>
-    where
-        S: AsRef<str>,
+        where
```
Comment (paulocsanz) [inline src/arraystring.rs:79]: Are you using `cargo fmt` to autoformat the code?

File: src/arraystring.rs [hunk 6]
```diff
@@ -649,12 +664,11 @@ impl<const N: usize> ArrayString<N> {
...
     pub fn push_str<S>(&mut self, string: S)
-    where
-        S: AsRef<str>,
+        where
+            S: AsRef<str>,
     {
         trace!("Push str truncate");
-        let size = Self::capacity().saturating_sub(self.len());
-        unsafe { self.push_str_unchecked(truncate_str(string.as_ref(), size)) }
+        let _ = self.try_push_str(truncate_str(string.as_ref(), self.remaining().len().into_lossy()));
```
Comment (paulocsanz) [inline src/arraystring.rs:671]: Why use try_push, that has an extra check when we already truncate it? Specially ignoring the error

**Comment by paulocsanz** [comment]
This is changing way too much critical stuff, thanks for the PR, but I will need time to ensure the correctness. The thing about the old code is that it was battle tested. So by changing everything we will require a lot of analysis.

I'm not against it, just informing you that I might be a bit annoying with the review.

**Comment by paulocsanz** [comment]
Also main is broken after that previous PR I will need some time to fix it. Mostly compilation when some features are disabled.

File: src/arraystring.rs [hunk 7]
```diff
@@ -712,10 +726,29 @@ impl<const N: usize> ArrayString<N> {
...
+    ///
+    /// // s = SmallString::try_from_str(&"0".repeat(SmallString::capacity().into()))?;
+    /// // Undefined behavior, don't do it
+    /// // s.push('!');
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[inline]
+    pub fn push(&mut self, ch: char) {
+        self.size = self.size.saturating_add(append_char_truncate(self.remaining_mut(), ch).into_lossy());
```
Comment (paulocsanz) [inline src/arraystring.rs:751]: Won't this break with ArrayString<255> with 255 as the length?

**Comment by paulocsanz** [comment]
Btw thank you for fixing the benchmark

**Comment by paulocsanz** [comment]
Thinking about it I'm ok with removing most unsafe usage even if performance degrades. At that point the user can depend on the unchecked functions and do the work themselves. It's way too hard to reason about unsafe, specially as it's copy-pasted everywhere.

File: src/utils.rs (hunk 2 above)
Comment (Wicpar) [reply src/utils.rs:142]: Yeah i should stop doing that, we can revert that line.

File: src/arraystring.rs (hunk 3 above)
Comment (Wicpar) [reply src/arraystring.rs:9]: Didn't really pay attention to that, i didn't realize std was a separate feature

File: src/arraystring.rs (hunk 5 above)
Comment (Wicpar) [reply src/arraystring.rs:79]: Might have forgotten to enable it

File: src/arraystring.rs (hunk 6 above)
Comment (Wicpar) [reply src/arraystring.rs:671]: It's to phase out the unchecked version. It's actually as fast as with the unchecked version except for smallstring where it has a 10% overhead (200 ps). You can see for yourself, i updated the benches to prevent parameter optimization.

File: src/arraystring.rs (hunk 7 above)
Comment (Wicpar) [reply src/arraystring.rs:751]: This is a checked version of the unchecked, it should be as performant as the unchecked one, but truncates automatically. We should add a bench to formalize that.

File: src/utils.rs (hunk 1 above)
Comment (Wicpar) [reply src/utils.rs:159]: In my tests the compiler will unroll 7 times which is too much. This version makes it a lot faster according to the benchmarks of push truncate, but we should do a comparative one specific to the function.

Is_char_boundary pretty much boils down to a range check and an and with zero check and micro-op fused short jump.
The only way i see to optimize it further is to operate directly on each byte unchecked. It is guaranteed to return even if we don't bound check on a valid utf8 str.

In theory the whole function could boil down to a branchless "magic" function that:
Get offset to start if < 4
Load unaligned 32 bits of the 4 previous bytes clamped to start
NAnd with magic value
Get trailing zeroes / 4 
Substract that value to the end of the slice clamped by the initial offset

Maybe one branch is more efficient where we do one where n >= 4 and the other where n < 4
For such short jumps that happen once there should be no performance impact from it.
Modern processors preload multiple branches, two should create no problem at all, where a while loop would.

File: src/arraystring.rs (hunk 6 above)
Comment (paulocsanz) [reply src/arraystring.rs:671]: Cool, I like this idea!

File: src/utils.rs (hunk 1 above)
Comment (paulocsanz) [reply src/utils.rs:159]: I was talking about replacing the while for a `for i in (0..4)` or something like that. But manually unrolling is a great improvement already.

Could you please open another PR, as this one got automatically closed when I renamed master to main?

Also, I would prefer if you could split the changes in the benchmark changes, and the optmizations/making things safer.

I would like to merge the benches, but I need more time to review all of the algo changes.
